test/smoke/webhook_signing_test.go

```diff
@@ -0,0 +1,85 @@
+package smoke
+
+import (
+	"encoding/base64"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/target/goalert/test/smoke/harness"
+)
+
+type WebhookTestingSign struct {
+	Body      []byte
+	Signature string
+}
+
+func TestWebhookCorrectlySigns(t *testing.T) {
+	t.Parallel()
+
+	ch := make(chan WebhookTestingSign, 1)
+
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		data, err := io.ReadAll(r.Body)
+		if !assert.NoError(t, err) {
+			return
+		}
+
+		ch <- WebhookTestingSign{
+			Body:      data,
+			Signature: r.Header.Get("X-Webhook-Signature"),
+		}
+	}))
+
+	defer ts.Close()
+
+	sql := `
+	insert into users (id, name, email) 
+	values 
+		({{uuid "user"}}, 'bob', 'joe');
+	insert into user_contact_methods (id, user_id, name, type, value) 
+	values
+		({{uuid "cm1"}}, {{uuid "user"}}, 'personal', 'WEBHOOK', '` + ts.URL + `');
+
+	insert into user_notification_rules (user_id, contact_method_id, delay_minutes) 
+	values
+		({{uuid "user"}}, {{uuid "cm1"}}, 0);
+
+	insert into escalation_policies (id, name) 
+	values
+		({{uuid "eid"}}, 'esc policy');
+	insert into escalation_policy_steps (id, escalation_policy_id) 
+	values
+		({{uuid "esid"}}, {{uuid "eid"}});
+	insert into escalation_policy_actions (escalation_policy_step_id, user_id) 
+	values 
+		({{uuid "esid"}}, {{uuid "user"}});
+
+	insert into services (id, escalation_policy_id, name, description) 
+	values
+		({{uuid "sid"}}, {{uuid "eid"}}, 'service', 'testing');
+
+	insert into alerts (service_id, summary, details, status, dedup_key) 
+	values
+		({{uuid "sid"}}, 'testing summary', 'testing details', 'triggered', 'auto:1:foo');
+`
+
+	h := harness.NewHarness(t, sql, "webhook-user-contact-method-type")
+	defer h.Close()
+
+	alert := <-ch
+
+	// convert alert.Signature from base64 to byte slice
+	signatureBytes, err := base64.StdEncoding.DecodeString(alert.Signature)
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	valid, _ := h.App().WebhookKeyring.Verify(alert.Body, signatureBytes)
+
+	if !assert.True(t, valid, "webhook signature invalid") {
+		return
+	}
```

Comment:
Adding the GraphQL/API bit (but not the UI changes) to grab the public key might be worthwhile so that the smoke test can demonstrate getting the key and verifying. 

It's also essential we _do NOT_ use any in-app code to verify the signature -- if the in-app keyring ever changes, the test won't detect the compatibility break; we need to insulate the two. This test should use it's own validation code, which should match whatever we'd document users use.

This will also help us understand how we expect verification to work externally.

Our existing keyring works well for internal GoAlert stuff. Still, we also need to do a little bit of digging into libraries of other languages to ensure validation isn't going to be painful if we use that too.

Reply:
I think the best method of ensuring external use-cases would be to allow the keyring to execute [`SignASN1`](https://pkg.go.dev/crypto/ecdsa@go1.23.4#SignASN1) and MAYBE [`VerifyASN1`](https://pkg.go.dev/crypto/ecdsa@go1.23.4#VerifyASN1) operations rather than sending `r`/`s` along side the signature itself; most libraries support ASN1 verification:

- [Java](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/security/Signature.html) via [BouncyCastle](https://downloads.bouncycastle.org/java/docs/bcprov-jdk18on-javadoc/org/bouncycastle/jcajce/provider/asymmetric/ec/SignatureSpi.ecDSA256.html) (unclear to me if it supports ECDSA with SHA512/224 digest as we currently use without some extra work)
- Python via [Cryptography](https://cryptography.io/en/latest/hazmat/primitives/asymmetric/ec/) (allows for precomputed hashes)
- Golang via crypto/ecdsa (requires digest input)
- Rust through RustCrypto or OpenSSL bindings (requires digest input)
- Erlang/Elixir through `:crypto` (can accept digest input)